### DISCUSSION TITLE
Discussion on Designing Apollo's Table Structure for Postgres: Join the Conversation

### DISCUSSION THREAD

**Discussion by hezhangjian** [apolloconfig/apollo#4772] [General]
Hi everyone, I'm shoothzj. Recently, I've been working on adapting Apollo to other relational databases, such as H2 and Postgres. I've designed a preliminary table structure for Apollo in the Postgres environment. Unlike H2, I think the database structure is more important in Postgres.

I'd like to organize a simple Chinese video conference to discuss this structure. The tentative time for the meeting is 8 PM on March 7th. I'll provide the link later. If anyone is interested, we can also discuss other related topics.

Attention：We will use chinese instead of english in the meeting

**Comment by nobodyiam** [comment]
There has been a lot of feature requests for Apollo to support databases other than MySQL, including Postgres, Oracle, etc. The support of in-memory databases such as h2 is also very useful for demo and quick start scenarios, so @Shoothzj's proposal is worthy of in-depth discussion. Whoever is interested in this topic is welcome to join.

**Comment by hezhangjian** [comment]
meeting link: https://meeting.huaweicloud.com/welink/#/j/99894159

**Comment by hezhangjian** [comment]
2023年3月7号Apollo讨论会议
Apollo Discussion Meeting on March 7, 2023

与会人：宋顺(nobodyiam)、Anonymous、hac、邹毅贤(zouyx)、王小权(Anilople)、倪思勇(nisiyong)、李钟豪、王乾、张、张乐、zhanglei、张赞赞、钟晓鹏、王乾、夏勇、张俭(shoothzj)

- 张俭分享了Apollo支持多数据库踩的一些坑和经验
- 会议讨论了支持多数据库的几个原则
  - 避免使用触发器的功能，像DataChanged_Time UPDATE，如果数据库支持可以用，不支持尽量在应用侧搞定
  - 尽量保持代码一致，不追求建表的数据库字段一致
  - 后续支持其他sql/nosql，优先考虑抽象storage层，可以以jpa接口不动的方式，书写其他数据（如mongo）的支持插件
- 关于apollo的参数校验、流控等问题，后续多和社区互动，形成合力
- 关于apollo openapi的集成问题，在社区open discussion一起讨论

- Zhang Jian shared some pitfalls and experience in supporting multiple databases in Apollo.
- The meeting discussed several principles for supporting multiple databases:
  - Avoid using trigger functions, such as DataChanged_Time UPDATE. If the database supports it, it can be used, otherwise, try to handle it on the application side.
  - Try to keep the code consistent, and do not pursue consistency in database table fields.
  - For supporting other SQL/NoSQL databases, prioritize abstracting the storage layer so that other data (such as MongoDB) can be written using JPA interfaces without modification.
- Regarding parameter validation and flow control issues in Apollo, there will be more interaction with the community to form a collective effort.
- Regarding the integration of Apollo OpenAPI, it will be discussed together in the community's open discussion.
